clients/wallet/src/task.rs

```diff
@@ -0,0 +1,206 @@
+	ProcessSuccess,
+	/// The error is acceptable, and is best to just retry again
+	RecoverableError,
+	/// Something happened when
+	Error(String),
+}
+
+pub struct LedgerTask {
+	ledger: Ledger,
+
+	/// receives the status of the task execution
+	task_execution_receiver: Receiver<LedgerTaskStatus>,
+
+	/// the number of retries allowed to process this specific task
+	retries_remaining: u8,
+
+	/// The latest status of this task
+	latest_status: LedgerTaskStatus,
+}
+
+impl LedgerTask {
+	/// Creates a task structure the `TransactionEnvelope`
+	pub fn create(ledger: Ledger, task_execution_receiver: Receiver<LedgerTaskStatus>) -> Self {
+		LedgerTask {
+			ledger,
+			task_execution_receiver,
+			// The decrement action is called everytime a new receiver is set.
+			retries_remaining: LEDGER_TASK_MAX_RETRIES,
+			latest_status: LedgerTaskStatus::NotStarted,
+		}
+	}
+
+	/// Sets a new one shot receiver.
+	/// Returns true if it was successful.
+	pub fn change_receiver(&mut self, receiver: Receiver<LedgerTaskStatus>) -> bool {
+		match self.latest_status {
+			LedgerTaskStatus::NotStarted | LedgerTaskStatus::RecoverableError
+				if self.retries_remaining > 0 =>
+			{
+				self.task_execution_receiver = receiver;
+				self.retries_remaining = self.retries_remaining.saturating_sub(1);
+				true
+			},
+			_ => false,
+		}
+	}
+
+	pub fn ledger(&self) -> Ledger {
+		self.ledger
+	}
+
+	/// Checks the status of the task
+	pub fn status(&mut self) -> LedgerTaskStatus {
+		if self.retries_remaining == 0 {
+			self.latest_status = LedgerTaskStatus::Error(format!(
+				"Transaction for ledger {} reached max limit of retries.",
+				self.ledger
+			));
+		} else {
+			self.latest_status = match &self.task_execution_receiver.try_recv() {
+				Ok(new_status) => match self.latest_status {
+					LedgerTaskStatus::NotStarted | LedgerTaskStatus::RecoverableError =>
+						new_status.clone(),
+					LedgerTaskStatus::ProcessSuccess | LedgerTaskStatus::Error(_) =>
+						self.latest_status.clone(),
+				},
+				Err(TryRecvError::Empty) => LedgerTaskStatus::NotStarted,
+				Err(TryRecvError::Closed) => LedgerTaskStatus::RecoverableError,
+			};
+		}
+
+		self.latest_status.clone()
+	}
+
+	/// Returns oneshot sender for sending status of a ledger
+	/// if and only if the status is `RecoverableError`
+	pub fn recover_and_create_sender(&mut self) -> Option<Sender<LedgerTaskStatus>> {
+		// Only recoverable errors can be given a new task.
+		if self.status() == LedgerTaskStatus::RecoverableError {
+			let (sender, receiver) = channel();
+
+			if self.change_receiver(receiver) {
+				self.latest_status = LedgerTaskStatus::NotStarted;
+				return Some(sender)
+			}
+		}
+
+		// For tasks flagged as `NotStarted` , wait for it.
+		// For tasks flagged as `ProcessSuccess`, they will be removed in the for loop
+		// For tasks flagged as `Error`, these will be removed.
+		// continue with the loop
+		None
+	}
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
It would be nice to have a test case that shows what happens once the maximum number of retries is exhausted and the 'recover' function is called.

Reply:
check test case `max_retries_exhausted()`